File: app/src/main/java/com/soya/launcher/utils/FileUtils.java
package com.soya.launcher.utils;

import android.content.Context;
import android.content.res.AssetManager;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.net.Uri;
import android.provider.MediaStore;
import android.util.Log;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;

/**
 * Created by ZTMIDGO on 2018/5/1.
 */

public class FileUtils {

    public static void copyAssets(AssetManager assetManager, String path, File outPath) throws IOException {
        String[] assets = assetManager.list(path);

        if (assets != null) {
            if (assets.length == 0) {
                copyFile(assetManager, path, outPath);
            } else {
                File dir = new File(outPath, path);
                if (!dir.exists()) {
                    if (!dir.mkdirs()) {
                        Log.v("copyAssets", "Failed to create directory " + dir.getAbsolutePath());
                    }
                }

                String[] var5 = assets;
                int var6 = assets.length;

                for(int var7 = 0; var7 < var6; ++var7) {
                    String asset = var5[var7];
                    copyAssets(assetManager, path + "/" + asset, outPath);
                }
            }

        }
    }

    private static void copyFile(AssetManager assetManager, String fileName, File outPath) throws IOException {
        Log.v("copyFile", "Copy " + fileName + " to " + outPath);
        InputStream in = assetManager.open(fileName);
        OutputStream out = new FileOutputStream(outPath + "/" + fileName);
        byte[] buffer = new byte[4000];

        int read;
        while((read = in.read(buffer)) != -1) {
            out.write(buffer, 0, read);
        }

        in.close();
        out.close();
    }

    public static void write(InputStream in, String filePath, String name) throws IOException {
        createPath(new File(filePath));
        int index;
        byte[] bytes = new byte[1024];
        OutputStream out = new FileOutputStream(filePath + "/" + name);
        while ((index = in.read(bytes)) != -1) {
            out.write(bytes, 0, index);
            out.flush();
        }
        in.close();
        out.close();
    }

    public static void createPath(File file){
        if (!file.exists()){
            file.mkdirs();
        }
    }

    public static String getFileSuffix(File file){
        return getFileSuffix(file.getName());
    }

    public static String getFileSuffix(String fileName){
        int index = fileName.lastIndexOf(".");
        return index == -1 ? null : fileName.substring(index);
    }

    public static String getFileNameWithoutSuffix(String fileName){
        return fileName.substring(0, fileName.lastIndexOf("."));
    }

    public static byte[] getBitmapBytes(Bitmap bitmap, Bitmap.CompressFormat format, int quality){
        try {
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            bitmap.compress(format, quality, baos);
            byte[] data = baos.toByteArray();
            baos.close();
            return data;
        }catch (Exception e){
            e.printStackTrace();
        }
        return null;
    }

    public static void deleteFile(String filePath){
        File file = new File(filePath);
        if (file.exists()){
            file.delete();
        }
    }

    public static void deleteAllFile(File file){
        if(file.isFile()){
            file.delete();
            return;
        }

        File[] files = file.listFiles();

        if(files == null)
            return;

        for(int i = 0; i < files.length; i++){
            File f = files[i];
            if(f.isFile()){
                f.delete();
            }else{
                deleteAllFile(f);
                f.delete();
            }
        }

        file.delete();
    }

    public static File renameFile(String filePath, String newName){
        File oldFile = new File(filePath);
        String basePath = oldFile.getParent();
        File newFile = new File(basePath + "/" +newName);
        oldFile.renameTo(newFile);
        return newFile;
    }

    public static File writeFile(byte[] data, String filePath, String fileName) throws IOException {
        createPath(new File(filePath));
        File file = new File(filePath+"/"+fileName);
        FileChannel fc = new FileOutputStream(file).getChannel();
        fc.write(ByteBuffer.wrap(data));
        fc.close();
        return file;
    }

    public static byte[] readStream(InputStream response) throws IOException {
        byte[] pool = new byte[1024];
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        int bytesRead = 0;
        try{
            while ((bytesRead = response.read(pool)) > 0){
                out.write(pool, 0, bytesRead);
            }
            return out.toByteArray();
        }catch (IOException e){
            e.printStackTrace();
        }finally {
            out.close();
            response.close();
        }
        return null;
    }

    public static byte[] readChannel(String filePath){
        try{
            FileChannel fc = new FileInputStream(filePath).getChannel();
            byte[] data = new byte[(int) fc.size()];
            fc.read(ByteBuffer.wrap(data));
            fc.close();
            return data;
        }
        catch (IOException e){
            e.printStackTrace();
        }
        return null;
    }

    public static byte[] readAssets(Context context, String name){
        try {
            InputStream in = context.getAssets().open(name);
            byte[] buffer = new byte[in.available()];
            in.read(buffer);
            in.close();
            return buffer;
        }catch (Exception e){
            e.printStackTrace();
        }
        return null;
    }

    public static void writeFile(InputStream in, String filePath) throws IOException {
        byte[] pool = new byte[1024];
        FileOutputStream out = new FileOutputStream(new File(filePath));
        int bytesRead = 0;
        try{
            while ((bytesRead = in.read(pool)) > 0){
                out.write(pool, 0, bytesRead);
            }
        }catch (IOException e){
            e.printStackTrace();
        }finally {
            out.close();
            in.close();
        }
    }

    public static String parseRealPathFromUri(Context context, Uri contentUri) {
        Cursor cursor = null;
        try {
            String[] proj = { MediaStore.Images.Media.DATA };
            cursor = context.getContentResolver().query(contentUri, proj, null, null, null);
            int column_index = cursor.getColumnIndex(MediaStore.Images.Media.DATA);
            cursor.moveToFirst();
            return cursor.getString(column_index);
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
    }
}

File: app/src/main/java/com/chihihx/launcher/utils/FileUtils.kt
package com.chihihx.launcher.utils

import android.content.Context
import android.content.res.AssetManager
import android.database.Cursor
import android.graphics.Bitmap
import android.graphics.Bitmap.CompressFormat
import android.net.Uri
import android.provider.MediaStore
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer

/**
 * Created by ZTMIDGO on 2018/5/1.
 */
object FileUtils {
    @Throws(IOException::class)
    fun copyAssets(assetManager: AssetManager, path: String, outPath: File) {
        val assets = assetManager.list(path)

        if (assets != null) {
            if (assets.size == 0) {
                copyFile(assetManager, path, outPath)
            } else {
                val dir = File(outPath, path)
                if (!dir.exists()) {
                    if (!dir.mkdirs()) {
                    }
                }

                val var5: Array<String> = assets
                val var6 = assets.size

                for (var7 in 0 until var6) {
                    val asset = var5[var7]
                    copyAssets(assetManager, "$path/$asset", outPath)
                }
            }
        }
    }

    @Throws(IOException::class)
    private fun copyFile(assetManager: AssetManager, fileName: String, outPath: File) {
        val `in` = assetManager.open(fileName)
        val out: OutputStream = FileOutputStream("$outPath/$fileName")
        val buffer = ByteArray(4000)

        var read: Int
        while ((`in`.read(buffer).also { read = it }) != -1) {
            out.write(buffer, 0, read)
        }

        `in`.close()
        out.close()
    }

    @Throws(IOException::class)
    fun write(`in`: InputStream, filePath: String, name: String) {
        createPath(File(filePath))
        var index: Int
        val bytes = ByteArray(1024)
        val out: OutputStream = FileOutputStream("$filePath/$name")
        while ((`in`.read(bytes).also { index = it }) != -1) {
            out.write(bytes, 0, index)
            out.flush()
        }
        `in`.close()
        out.close()
    }

    fun createPath(file: File) {
        if (!file.exists()) {
            file.mkdirs()
        }
    }

    fun getFileSuffix(file: File): String? {
        return getFileSuffix(file.name)
    }

    fun getFileSuffix(fileName: String): String? {
        val index = fileName.lastIndexOf(".")
        return if (index == -1) null else fileName.substring(index)
    }

    fun getFileNameWithoutSuffix(fileName: String): String {
        return fileName.substring(0, fileName.lastIndexOf("."))
    }

    fun getBitmapBytes(bitmap: Bitmap, format: CompressFormat?, quality: Int): ByteArray? {
        try {
            val baos = ByteArrayOutputStream()
            bitmap.compress(format!!, quality, baos)
            val data = baos.toByteArray()
            baos.close()
            return data
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    fun deleteFile(filePath: String?) {
        val file = File(filePath)
        if (file.exists()) {
            file.delete()
        }
    }

    fun deleteAllFile(file: File) {
        if (file.isFile) {
            file.delete()
            return
        }

        val files = file.listFiles() ?: return

        for (i in files.indices) {
            val f = files[i]
            if (f.isFile) {
                f.delete()
            } else {
                deleteAllFile(f)
                f.delete()
            }
        }

        file.delete()
    }

    fun renameFile(filePath: String?, newName: String): File {
        val oldFile = File(filePath)
        val basePath = oldFile.parent
        val newFile = File("$basePath/$newName")
        oldFile.renameTo(newFile)
        return newFile
    }

    @Throws(IOException::class)
    fun writeFile(data: ByteArray?, filePath: String, fileName: String): File {
        createPath(File(filePath))
        val file = File("$filePath/$fileName")
        val fc = FileOutputStream(file).channel
        fc.write(ByteBuffer.wrap(data))
        fc.close()
        return file
    }

    @Throws(IOException::class)
    fun readStream(response: InputStream): ByteArray? {
        val pool = ByteArray(1024)
        val out = ByteArrayOutputStream()
        var bytesRead = 0
        try {
            while ((response.read(pool).also { bytesRead = it }) > 0) {
                out.write(pool, 0, bytesRead)
            }
            return out.toByteArray()
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            out.close()
            response.close()
        }
        return null
    }

    fun readChannel(filePath: String?): ByteArray? {
        try {
            val fc = FileInputStream(filePath).channel
            val data = ByteArray(fc.size().toInt())
            fc.read(ByteBuffer.wrap(data))
            fc.close()
            return data
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return null
    }

    fun readAssets(context: Context, name: String?): ByteArray? {
        try {
            val `in` = context.assets.open(name!!)
            val buffer = ByteArray(`in`.available())
            `in`.read(buffer)
            `in`.close()
            return buffer
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    @Throws(IOException::class)
    fun writeFile(`in`: InputStream, filePath: String?) {
        val pool = ByteArray(1024)
        val out = FileOutputStream(File(filePath))
        var bytesRead = 0
        try {
            while ((`in`.read(pool).also { bytesRead = it }) > 0) {
                out.write(pool, 0, bytesRead)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            out.close()
            `in`.close()
        }
    }

    fun parseRealPathFromUri(context: Context, contentUri: Uri?): String {
        var cursor: Cursor? = null
        try {
            val proj = arrayOf(MediaStore.Images.Media.DATA)
            cursor = context.contentResolver.query(contentUri!!, proj, null, null, null)
            val column_index = cursor!!.getColumnIndex(MediaStore.Images.Media.DATA)
            cursor.moveToFirst()
            return cursor.getString(column_index)
        } finally {
            cursor?.close()
        }
    }
}
